inversion-cosmosdb/src/main/java/io/inversion/cosmosdb/CosmosDb.java

```diff
@@ -151,7 +151,31 @@ public String upsertRow(Collection collection, Map<String, Object> row) throws A
             if (id == null) {
                 id = collection.encodeResourceKey(row);
                 if (id == null)
+                {
+                    // this is the first potential problem, where we couldn't get the primary index for some reason
+                    Index index = collection.getPrimaryIndex();
+                    if(index == null) {
+                        log.warn("INVERSION UNABLE TO DETERMINE PRIMARY INDEX");
+                        collection.getIndexes().forEach(cIndex -> {
+                            log.info("    {}", cIndex.toString());
```

Comment:
toString is extraneous

Reply:
The index listing is only going to happen on an error.  I would like to see a single string logged on warn only.  That goes for the bottom debug too.  I appreciate the thinking here about fine tuning this stuff but I have started to think about finding all of your applicable logging for a single event in something like CloudWatch where you could be sifting through hundreds of concurrent events.  In that type of concurrent environment there is nothing that obviously links the line 158 error to the line 160 information in the log data itself.  If the spatial sequence is lost you have to look harder.  Not worth it.  Just go a single warn call w/ a big string.

Reply:
In the case of Azure, the logs for a particular thread are all joined on a common operation ID. So all I have to do is to find an instance of the all caps error, get its operation id, and then get all logs for that operation id to see the rest of the messages for that thread. If this is simply a matter of being to chatty I can reduce the number of log lines, but finding and correlating them isn't an issue.

And yeah, I got into the habit years (maybe even decades) ago of always toString()ing shit because it didn't always at as I expected. Probably not necessary in the 21s century! :)

Reply:
Gotta think beyond your awesome azure app insights skills! Think a little bit lowest common denominator here :)

Reply:
However, it does occur to me that we don't need to list the indexes here, we have them in the API config. We just need to know that the index couldn't be determined.